### DISCUSSION TITLE
Closures and higer order functions in mojo

### DISCUSSION THREAD

**Discussion by ekinimo** [modular/modular#244] [General]
I couldnt figure out how to  make a closure in mojo. I wonder if it is supported, or whether there are plans to support it in the future.
i just hope this will result in something more sensible than pythons result
```python
lambdas = []
for i in range(10):
   lambdas.append(lambda x: x+i)
print(lambdas[3](0))
```

**Comment by ksandvik** [comment]
If this is missing, could we have labels for missing (and to be implemented) Python features so it's easy to find them in the GitHub issues database?

**Comment by daydun** [comment]
Mojo does have closures, just not the `lambda` syntax yet.
```py
from Vector import DynamicVector

var lambdas = DynamicVector[fn(Int) capturing -> Int]()
for i in range(10):
    fn f(x: Int) -> Int:
        return x + i
    lambdas.push_back(f)
print(lambdas[3](0))
```

But be careful about https://github.com/modularml/mojo/issues/217

**Reply by ekinimo** [reply]
So this is basicly analogous to [gcc s nested functions ?](https://gcc.gnu.org/onlinedocs/gcc/Nested-Functions.html)

**Reply by ekinimo** [reply]
Also given code does what i expect but not what python does, if i change `fn` to `def`  it fails. I think this behaviour is perfectly sensible but not compatible with python lambdas  (the same code would result differently). Is this to be adressed in the future ? What should i expect ?

**Reply by daydun** [reply]
I am not familiar with gcc nested functions. Closures in mojo work like I'd expect closures to usually work, capturing referenced variables if required.

Switching to `def` means `f` can now also raise errors, which appears to cause issues https://github.com/modularml/mojo/issues/255.

**Reply by survuvi** [reply]
@DayDun on your code how to print the full vector, so print all elements instead of only third indexed element?

```python
from Vector import DynamicVector

var lambdas = DynamicVector[fn(Int) capturing -> Int]()
for i in range(10):
    fn f(x: Int) -> Int:
        return x + i
    lambdas.push_back(f)
print(*lambdas) #error: Expression [4]:23:12: unpacked arguments are not supported yet
```

**Reply by daydun** [reply]
With a loop

**Comment by survuvi** [comment]
@ekinimo 

```python

%%python
lambdas = []
for i in range(10):
   lambdas.append(lambda x: x+i)
print(lambdas[3](0)) # Output: 9, due to the late binding closure issue

```

```python

%%python
lambdas = []
for i in range(10):
    lambdas.append(lambda x, i=i: x+i)
print(lambdas[3](0))  # Output: 3

```

**Reply by daydun** [reply]
Yeah no clue how mojo will handle this, will be interesting to see.

**Reply by survuvi** [reply]
at moment I'm crazy with this closure:

```python

fn main(n : Int) -> Int:
    @parameter
    fn multiplication_table() -> Int:
        var i_1 : Int = 1
        while i_1 < 10:
            print(i_1*n)
        return i_1
    return n

main(5)

```

_warning: Expression [46]:8:9: 'i_1' was declared as a 'var' but never mutated, consider switching to a 'let'
        var i_1 : Int = 1_

I tried also:

```python

fn main(n : Int) -> fn multiplication_table():
    var i_1 : Int = 1
    while i_1 < 10:
        print(i_1*n)
    return i_1
return n

main(5)

```

_error: Expression [49]:5:24: expected '(' for argument list
fn main(n : Int) -> fn multiplication_table():

error: Expression [49]:5:24: use of unknown declaration 'multiplication_table', 'fn' declarations require explicit variable declarations
fn main(n : Int) -> fn multiplication_table():

error: Expression [49]:21:12: use of unknown declaration 'n'
    return n_

**Reply by lattner** [reply]
Do you mean to increment i_1 in the loop?

**Reply by survuvi** [reply]
yes @lattner the output should be a multiplication table:
_5
10
15
20
25
30
35
40
45
50_

**Reply by lattner** [reply]
Ok, you need a `+=` in there somewhere, e.g.:

```
 var i_1 : Int = 1
        while i_1 < 10:
            print(i_1*n)
            i_1 += 1
        return i_1
```

because 'while' doesn't increment the induction variable

**Reply by survuvi** [reply]
Thank you @lattner could be there is a bug bacause i_1*n not print anything

```python

fn main(n : Int) -> Int:
    @parameter
    fn multiplication_table() -> Int:
         var i_1 : Int = 1
        while i_1 < 10:
            print(i_1*n)
            i_1 += 1
        return i_1
    return n

main(5)

```

**Reply by lattner** [reply]
You also need to call the `multiplication_table()` function :-)

**Reply by survuvi** [reply]
```python

fn main(n : Int) -> Int:
    @parameter
    fn multiplication_table() -> Int:
        var i_1 : Int = 1
        while i_1 < 10:
            print(i_1*n)
            i_1 += 1
        return i_1
    return n

let func = main(5)

var result = func()

print(result)

print(main(5)())

```

_error: Expression [11]:28:22: 'Int' does not implement the '__call__' method
    var result = func()
                     ^

error: Expression [11]:33:18: 'Int' does not implement the '__call__' method
    print(main(5)())_

**Reply by lattner** [reply]
https://github.com/modularml/mojo/issues/272#issuecomment-1565767376

**Comment by lattner** [comment]
Loosely held opinion, Mojo clearly needs to support:

1) Nested functions (currently wired up, but have a few issues given lifetimes are not here yet).  I'd like `@parameter` to go away on the nested functions eventually too.
2) Existing Python lambda syntax, which is sugar for #1.  We need to support type annotations here.

Lower priority, but I think we're likely to explore:

3) Possibly implement more flexible/general/ergonomic light-weight closures like Scala3 `=>` syntax
4) User defined statement blocks, e.g.:

```
  parallel_loop(42):
    stuff()
```

User defined statements are a nice way to shift more language syntax into the library, but are just syntactic sugar and will require a little more infra to get wired up. For example, I would like "return" in that context to return from the enclosing function (not from the lambda), and things like break to work for loop-like constructs.  This is quite possible to wire up, but will require a bit of design work.

**Reply by daydun** [reply]
User defined statements would be epic. I love the idea of moving things from the compiler to libraries, but you have to of course be careful to not do it at the expense of other things.

I think it's very neat for example how Swift defines the short-circuiting `&&` operator completely in the [standard library](https://github.com/apple/swift/blob/495f9b210507305c65c5064c333efd32126c06e7/stdlib/public/core/Bool.swift#L280).

**Reply by lattner** [reply]
Yep, lots of lessons there.  It still bugs me how "return" works the wrong way and break doesn't work in a "closure taking control flow" function in Swift.  We can do better.

**Reply by nmsmith** [reply]
Chris, one thing I would love ❤️ 😍 is to have a version of these user-defined statement blocks that supports arbitrary Unicode text within them. Mojo's parser would strip the leading indentation, tokenize the text, and pass the tokens to the block's implementation. Additionally, Mojo values could be interpolated using Python's f-string syntax `{}`, or something akin to it.

What I'm describing is basically a multi-line template literal, delimited by indentation, rather than `"` characters. I'm also proposing to pass the implementation a list of pre-tokenized lines, rather than a raw string, so that nobody has to worry about weird whitespace issues.

Library authors could use this feature to embed DSLs within Mojo, and the DSL programs can be compiled at the same time as Mojo using Mojo's metaprogramming feature, allowing them to run lightning fast. The end result could be something **epic!**

(A lot of details would need to be fleshed out, of course.)

**Reply by lattner** [reply]
You might have noticed that I'm a fan of powerful library-defined metaprogramming features :-).  I look forward to exploring all this, but we must stay focused :-)
